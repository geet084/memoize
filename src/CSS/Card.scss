.card {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
}

.count {
  color: #00AEFF;
}

.definition {
  width: 75%;
  height: 110px;
  margin: 5px;
  padding: 10px;
}

.num-guess {
  width: 40%;
  color: #fff;
  background-color: #016FB9;
  margin: 20px 5px 5px 5px;
  padding: 10px;
  border-radius: 15px;
}

.show-previous {
  width: 40%;
  margin: 5px;
  padding: 10px;
  color: #fff;
  background-color: #016FB9;
  border-radius: 15px;
}
.undefined {
  display: none;
}

.yes {
  color: lightgreen;
}

.no {
  color: maroon;
}

.user-btn {
  width: 290px;
  height: 35px;
  margin: 10px;
  border-radius: 10px;
  font-size: 1rem;
  background-color: #00AEFF;
  &:hover {
    color: white;
    background-color: #2D4C46;
  }
}

.user-input {
  width: 285px;
  height: 30px;
  margin: 10px;
  padding: 2px 2px 2px 5px;
  font-size: 1.25rem;
  background-color: #00AEFF;
  &::placeholder {
    color:white;
  }
  &:hover {
    color: white;
    background-color: #2D4C46;
  }
}

