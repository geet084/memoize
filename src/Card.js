import React, { Component } from 'react';
import './CSS/Main.scss';

export default class Card extends Component {
  constructor(props) {
    super(props);
    
    this.state = {
      count: 0,
      question: null,
      answeredCorrectly: null,
      newStart: true
    }
  }

  componentDidMount = () => {
    this.showQuestion();
  }
  
  begin = () => {
    this.setState({newStart: false})
  }

  getRandomNumber = (max) => {
    return Math.floor(Math.random() * max)
  }

  showQuestion = () => {
    let { prototypes, selection } = this.props.deets;
    let category = prototypes[selection];
    let randomIndex = this.getRandomNumber(category.length)

    let rightAnswer = this.buildCard('keys', category, randomIndex).shift()
    let definition = this.buildCard('values', category, randomIndex).shift()
    // let wrongAnswers = this.generateWrongAnswers(category, randomIndex)

    this.setState({ question: {question: definition, answer: rightAnswer} });
    // this.setState({ question: [definition, [rightAnswer, ...wrongAnswers].sort()] });
  }

  buildCard = (type, category, index) => {
    if (type === 'keys') {
      return Object.keys(category[index])
    } else if (type === 'values') {
      return Object.values(category[index])
    }
  }

  // generateWrongAnswers = (category, random) => {
  //   let i = 0
  //   let arr = []
 
  //   while (i < 3) {
  //     let j = this.getRandomNumber(category.length);
  //     if (j !== random) {
  //       arr.push(Object.keys(category[j]).shift());
  //       i++;
  //     } 
  //   };

  //   return arr;
  // }
  
  // checkAnswer = (event) => {
  //   let guess = event.target.innerText;
  //   let correct = false;
    
  //   if (guess === this.state.question[1][0]) correct = true;
  //   this.props.answer(correct, this.state.question)
  //   this.setState({
  //     count: this.state.count + 1,
  //     answeredCorrectly: correct
  //   })
  //   this.showQuestion();
  // }

  checkAnswer = (e) => {
    let guess = e.target.innerText.toLowerCase();
    let { rightAnswer } = this.state.question.answer;
    let correct = false;

    if (guess === rightAnswer.toLowerCase()) {
      correct = true;
    }

    this.props.answer(correct, this.state.question)
    this.setState({
      count: this.state.count + 1,
      answeredCorrectly: correct
    })
    this.showQuestion();
  }
  
  render() {
    // let { index, searchFound } = this.props;  
    // let { prototypes } = this.props.deets;
    // let { all } = this.props.deets.prototypes;
    // <p>{this.buildCard('keys', all, index).shift() + '()'}</p>
    // <p>{this.buildCard('values', all, index).shift()}</p>

   if (this.state.newStart) {
      return (
        <div>
          <h4 onClick={this.begin}>Click to begin</h4>
        </div>
      )
   } else {
     let { question, answer } = this.state.question;
     console.log(this.state.question)
      return (
        <div>
          <p>Questions answered so far: {this.state.count}</p>
          <p>{question} //{answer}</p>
          {/* <ul>
            {
              answers.map((answer, index) => {
                return <li onClick={this.checkAnswer}
                           key={index}>{answer}</li>
              })
            }
          </ul> */}
          <input type="text" onClick={this.checkAnswer}/>
        </div>
      )
    }
  }
}