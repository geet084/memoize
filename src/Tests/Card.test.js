import React from 'react';
import Card from '../Card.js';
import { shallow } from 'enzyme';
import aTypes from '../datasets/aType.js'

const mockPrototypes = jest.fn();
const mockAnswers = jest.fn();
const mockShowQuestion = jest.fn();
const mockSingleQuestion = [
  {
    "concat": "This method is used to merge two or more arrays. This method also does not change the existing arrays, but instead returns a new array."
  }]
const mockAnsweredQuestions = [{
  question: false,
  guess: { answer: 'push', definition: "This method adds one or more elements to the end of an array and returns the new length of the array."}
}];

describe('Card', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(
      <Card questions={mockPrototypes}
        answer={mockAnswers} />
    )
    expect(wrapper.state()).toEqual({
      count: 0,
      question: null,
      answeredCorrectly: null,
      showNextQuestion: false
    })
  })

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  })

  it('should have a proper default state', () => {
    expect(wrapper.state()).toEqual({
      count: 0,
      question: null,
      answeredCorrectly: null,
      showNextQuestion: false
    })
  })

  it.skip('should advance to the next question when clicked', () => {
    wrapper.setState({
      showNextQuestion: false,
      question: {
        definition: aTypes.aTypes[6].find,
        answer: 'find'
      }, 
      count: 1,
      answeredCorrectly: true
    })
    wrapper.setProps({answeredQuestions: mockAnsweredQuestions})
    wrapper.instance().nextQuestion();
    // expect(mockShowQuestion).toBeCalled()

    expect(wrapper.state()).toEqual({
      showNextQuestion: true,
      count: 0,
      answeredCorrectly: null
    })
  })
  
  it('should generate a random number between zero and a given max', () => {
    // RANDOM NUMBER...PASSES MOST OF THE TIME??
    let number = wrapper.instance().getRandomNumber(10)

    expect(number).not.toEqual(10)
  })

  it('should show a question when prompted', () => {
  /*SPECIFIC TO NOT HAVE TO DEAL WITH THE RANDOM GENERATED PART!!*/
    
    wrapper.setProps({questions: mockSingleQuestion})
    wrapper.instance().showQuestion();
    expect(wrapper.state()).toEqual({
      count: 0,
      question: {
        answer: 'concat',
        definition: 'This method is used to merge two or more arrays. This method also does not change the existing arrays, but instead returns a new array.'
      },
      answeredCorrectly: null,
      showNextQuestion: false
    })
  })

  it.skip('should find a question and answer from available choices', () => {
    let mockType = 'keys';
    let mockIndex = 1;
    wrapper.setProps({ questions: mockSingleQuestion })
    expect(wrapper.instance().findQuestion(mockType, mockIndex)).toEqual('copyWithin')
    
    mockType = 'values';
  
    expect(wrapper.instance().findQuestion(mockType, mockIndex)).toEqual('This method shallow copies part of an array to another location in the same array and returns it, without modifying its size.')
  })

  it('should check the answer that the user puts in', () => {

    expect(wrapper.state()).toEqual({
      count: 0,
      question: null,
      answeredCorrectly: null,
      showNextQuestion: false
    })
    
    wrapper.setState({
      question: {
        definition: aTypes.aTypes[6].find,
        answer: 'find'
      }
    })
    let mockClick = { target: { previousSibling: { value: 'hello' } } }
    wrapper.instance().checkAnswer(mockClick);

    expect(wrapper.state()).toEqual({
      count: 1,
      question: {
        definition: aTypes.aTypes[6].find,
        answer: 'find'
      },
      answeredCorrectly: false,
      showNextQuestion: false
    })
  })

  it.skip('should so the results of previous correctly answered questions', () => {
    expect(wrapper.state()).toEqual({
      count: 0,
      question: null,
      answeredCorrectly: null,
      showNextQuestion: false
    })
    wrapper.setProps({ answeredQuestions: mockSingleQuestion })
    let result = wrapper.instance().showPrevResult();
    expect(result).toEqual('undefined')
  })

  it('should toggle the check answer button text', () => {
    let result = wrapper.instance().showBtnText();
    expect(result).toEqual('Click to check answer')
    wrapper.setState({ count: 9 })
    result = wrapper.instance().showBtnText();
    expect(result).toEqual('Click to try again')
  })
});

