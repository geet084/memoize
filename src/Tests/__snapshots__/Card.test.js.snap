// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot with all data passed in correctly 1`] = `
<div>
  <h4
    className="begin"
    onClick={[Function]}
  >
    Click to begin
  </h4>
  <p
    className="intro"
  >
    Welcome to Study Time, a web-based flashcard app to practice how well you know your Javascript Array Prototypes! 
  </p>
</div>
`;
